From: Gavin Lambert <gavin.lambert@tomra.com>
Date: Wed, 1 Jun 2016 10:01:51 +0200

It's important to still let the watchdog run even when receiving packets, or the
link detection misbehaves when using master-level redundancy.

Also the watchdog task needs to run on a background thread instead of the RT
thread, to reduce latency spikes.

This has been reported to cause issues when using RTAI applications; consider
not applying it when using RTAI.

diff --git a/devices/e1000e/netdev-2.6.32-ethercat.c b/devices/e1000e/netdev-2.6.32-ethercat.c
--- a/devices/e1000e/netdev-2.6.32-ethercat.c
+++ b/devices/e1000e/netdev-2.6.32-ethercat.c
@@ -551,7 +551,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,status,rx_desc->special);
 		}
@@ -878,7 +877,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1068,7 +1066,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, status,
 			                  rx_desc->special);
diff --git a/devices/e1000e/netdev-2.6.33-ethercat.c b/devices/e1000e/netdev-2.6.33-ethercat.c
--- a/devices/e1000e/netdev-2.6.33-ethercat.c
+++ b/devices/e1000e/netdev-2.6.33-ethercat.c
@@ -519,7 +519,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,status,rx_desc->special);
 		}
@@ -879,7 +878,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1069,7 +1067,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, status,
 			                  rx_desc->special);
diff --git a/devices/e1000e/netdev-2.6.34-ethercat.c b/devices/e1000e/netdev-2.6.34-ethercat.c
--- a/devices/e1000e/netdev-2.6.34-ethercat.c
+++ b/devices/e1000e/netdev-2.6.34-ethercat.c
@@ -520,7 +520,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,status,rx_desc->special);
 		}
@@ -877,7 +876,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1067,7 +1065,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, status,
 			                  rx_desc->special);
diff --git a/devices/e1000e/netdev-2.6.35-ethercat.c b/devices/e1000e/netdev-2.6.35-ethercat.c
--- a/devices/e1000e/netdev-2.6.35-ethercat.c
+++ b/devices/e1000e/netdev-2.6.35-ethercat.c
@@ -869,7 +869,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,status,rx_desc->special);
 		}
@@ -1221,7 +1220,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1412,7 +1410,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, status,
 			                  rx_desc->special);
diff --git a/devices/e1000e/netdev-2.6.37-ethercat.c b/devices/e1000e/netdev-2.6.37-ethercat.c
--- a/devices/e1000e/netdev-2.6.37-ethercat.c
+++ b/devices/e1000e/netdev-2.6.37-ethercat.c
@@ -877,7 +877,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,status,rx_desc->special);
 		}
@@ -1226,7 +1225,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1417,7 +1415,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, status,
 			                  rx_desc->special);
@@ -5769,7 +5766,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.10-ethercat.c b/devices/e1000e/netdev-3.10-ethercat.c
--- a/devices/e1000e/netdev-3.10-ethercat.c
+++ b/devices/e1000e/netdev-3.10-ethercat.c
@@ -1022,7 +1022,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1450,7 +1449,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1641,7 +1639,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6623,7 +6620,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.12-ethercat.c b/devices/e1000e/netdev-3.12-ethercat.c
--- a/devices/e1000e/netdev-3.12-ethercat.c
+++ b/devices/e1000e/netdev-3.12-ethercat.c
@@ -1020,7 +1020,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1448,7 +1447,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1639,7 +1637,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6675,7 +6672,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.14-ethercat.c b/devices/e1000e/netdev-3.14-ethercat.c
--- a/devices/e1000e/netdev-3.14-ethercat.c
+++ b/devices/e1000e/netdev-3.14-ethercat.c
@@ -1020,7 +1020,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1448,7 +1447,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1639,7 +1637,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6683,7 +6680,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.16-ethercat.c b/devices/e1000e/netdev-3.16-ethercat.c
--- a/devices/e1000e/netdev-3.16-ethercat.c
+++ b/devices/e1000e/netdev-3.16-ethercat.c
@@ -1045,7 +1045,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1488,7 +1487,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1679,7 +1677,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6829,7 +6826,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.18-ethercat.c b/devices/e1000e/netdev-3.18-ethercat.c
--- a/devices/e1000e/netdev-3.18-ethercat.c
+++ b/devices/e1000e/netdev-3.18-ethercat.c
@@ -1045,7 +1045,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1488,7 +1487,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1679,7 +1677,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6857,7 +6854,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.2-ethercat.c b/devices/e1000e/netdev-3.2-ethercat.c
--- a/devices/e1000e/netdev-3.2-ethercat.c
+++ b/devices/e1000e/netdev-3.2-ethercat.c
@@ -982,7 +982,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1354,7 +1353,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1541,7 +1539,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6051,7 +6048,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.4-ethercat.c b/devices/e1000e/netdev-3.4-ethercat.c
--- a/devices/e1000e/netdev-3.4-ethercat.c
+++ b/devices/e1000e/netdev-3.4-ethercat.c
@@ -977,7 +977,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1381,7 +1380,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1567,7 +1565,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6205,7 +6202,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.6-ethercat.c b/devices/e1000e/netdev-3.6-ethercat.c
--- a/devices/e1000e/netdev-3.6-ethercat.c
+++ b/devices/e1000e/netdev-3.6-ethercat.c
@@ -969,7 +969,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1378,7 +1377,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1564,7 +1562,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6224,7 +6221,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-3.8-ethercat.c b/devices/e1000e/netdev-3.8-ethercat.c
--- a/devices/e1000e/netdev-3.8-ethercat.c
+++ b/devices/e1000e/netdev-3.8-ethercat.c
@@ -968,7 +968,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1371,7 +1370,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb,
 					  staterr, rx_desc->wb.middle.vlan);
@@ -1560,7 +1558,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6162,7 +6159,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-4.1-ethercat.c b/devices/e1000e/netdev-4.1-ethercat.c
--- a/devices/e1000e/netdev-4.1-ethercat.c
+++ b/devices/e1000e/netdev-4.1-ethercat.c
@@ -1046,7 +1046,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1489,7 +1488,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1680,7 +1678,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -6879,7 +6876,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-4.14-ethercat.c b/devices/e1000e/netdev-4.14-ethercat.c
--- a/devices/e1000e/netdev-4.14-ethercat.c
+++ b/devices/e1000e/netdev-4.14-ethercat.c
@@ -1047,7 +1047,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1496,7 +1495,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1687,7 +1685,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -7173,7 +7170,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-4.19-ethercat.c b/devices/e1000e/netdev-4.19-ethercat.c
--- a/devices/e1000e/netdev-4.19-ethercat.c
+++ b/devices/e1000e/netdev-4.19-ethercat.c
@@ -1027,7 +1027,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1480,7 +1479,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1671,7 +1669,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -7151,7 +7148,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog(&adapter->watchdog_timer);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-4.4-ethercat.c b/devices/e1000e/netdev-4.4-ethercat.c
--- a/devices/e1000e/netdev-4.4-ethercat.c
+++ b/devices/e1000e/netdev-4.4-ethercat.c
@@ -1046,7 +1046,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1489,7 +1488,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1680,7 +1678,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -7105,7 +7102,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
diff --git a/devices/e1000e/netdev-4.9-ethercat.c b/devices/e1000e/netdev-4.9-ethercat.c
--- a/devices/e1000e/netdev-4.9-ethercat.c
+++ b/devices/e1000e/netdev-4.9-ethercat.c
@@ -1046,7 +1046,6 @@ static bool e1000_clean_rx_irq(struct e1
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 		    e1000_receive_skb(adapter, netdev, skb, staterr,
 				      rx_desc->wb.upper.vlan);
@@ -1495,7 +1494,6 @@ copydone:
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					rx_desc->wb.middle.vlan);
@@ -1686,7 +1684,6 @@ static bool e1000_clean_jumbo_rx_irq(str
 
 		if (adapter->ecdev) {
 			ecdev_receive(adapter->ecdev, skb->data, length);
-			adapter->ec_watchdog_jiffies = jiffies;
 		} else {
 			e1000_receive_skb(adapter, netdev, skb, staterr,
 					  rx_desc->wb.upper.vlan);
@@ -7165,7 +7162,7 @@ void ec_poll(struct net_device *netdev)
 	if (jiffies - adapter->ec_watchdog_jiffies >= 2 * HZ) {
 		struct e1000_hw *hw = &adapter->hw;
 		hw->mac.get_link_status = true;
-		e1000_watchdog_task(&adapter->watchdog_task);
+		e1000_watchdog((unsigned long) adapter);
 		adapter->ec_watchdog_jiffies = jiffies;
 	}
 
